# For Review; please check if there are easyier faster ways 

#### Your tasks in this project are as follows:

- Data wrangling, which consists of:
    - Gathering data (downloadable file in the Resources tab in the left most panel of your classroom and linked in step 1 below).
    - Assessing data
    - Cleaning data
- Storing, analyzing, and visualizing your wrangled data
- Reporting on 1) your data wrangling efforts and 2) your data analyses and visualizations

##### Storing, Analyzing, and Visualizing Data for this Project
Store the clean DataFrame(s) in a CSV file with the main one named twitter_archive_master.csv. If additional files exist because multiple tables are required for tidiness, name these files appropriately. Additionally, you may store the cleaned data in a SQLite database (which is to be submitted as well if you do).

Analyze and visualize your wrangled data in your wrangle_act.ipynb Jupyter Notebook. At least three (3) insights and one (1) visualization must be produced.

##### Reporting for this Project
Create a 300-600 word written report called wrangle_report.pdf or wrangle_report.html that briefly describes your wrangling efforts. This is to be framed as an internal document.

Create a 250-word-minimum written report called act_report.pdf or act_report.html that communicates the insights and displays the visualization(s) produced from your wrangled data. This is to be framed as an external document, like a blog post or magazine article, for example.

Both of these documents can be created in separate Jupyter Notebooks using the Markdown functionality of Jupyter Notebooks, then downloading those notebooks as PDF files or HTML files (see image below). You might prefer to use a word processor like Google Docs or Microsoft Word, however.

# Wrangle and Analyze Data: ...

This project ....

## Table of Contents
- [0. Introduction](#intro)
- [1. Gather data](#gather)
- [2. Assess data](#assess)
- [3. Clean data](#clean)
- [4. Analysis & Visualization](#analysis)


<a id='gather'></a>
## 0. Introduction

bli bla bluo

In [767]:
# Import necessary libraries
import numpy as np
import pandas as pd
import requests # to download files programmatically 
import os # to work with local directory
import tweepy # for twitter-api
import time # for timer 
import json # to create json file from python dictionary

<a id='intro'></a>
## 1. Gather data

####  Data is gathered from three different sources of data as described in steps below:

1. The WeRateDogs Twitter archive. File `twitter_archive_enhanced.csv` as comma separated file in local directory.
2. The tweet image predictions, i.e., what breed of dog (or other object, animal, etc.) is present in each tweet according to a neural network. This file (image_predictions.tsv) is hosted on Udacity's servers at [URL](https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv).
3. Each tweet's retweet count and favorite ("like") count. Using the tweet IDs in the WeRateDogs Twitter archive,  the Twitter API is queried for each tweet's JSON data using Python's Tweepy library and each tweet's entire set of JSON data is stored in a file called tweet_json.txt file. Each tweet's JSON data is written to its own line. Then this .txt file is read line by line into a pandas DataFrame with tweet ID, retweet count, and favorite count.

### a. Read data from csv-file

In [490]:
# Read WeRateDogs Twitter archive from csv
df_twitter_archive_enhanced = pd.read_csv('twitter-archive-enhanced.csv')

### b. Programmatically download file from url

In [491]:
# Create request-object with url of file
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)

In [492]:
# Create file with content of response-object
with open(url.split('/')[-1], mode='wb') as file: 
    file.write(response.content)

In [493]:
# Read tweet image predictions from tsv
df_image_predictions = pd.read_csv(url.split('/')[-1], sep='\t')

### c. Query additional information for tweets via Twitter API

In [494]:
# Define credentials for twitter api
try:
    with open("pw.json", "r") as read_file:
        pw = json.load(read_file)

# Catch exception, if file doesn't exist
except FileNotFoundError as fnf_error:
    print(fnf_error)
    print('File pw.json with twitter credentials missing')
    
consumer_key = pw['consumer_key']
consumer_secret = pw['consumer_secret']
access_token = pw['access_token']
access_secret = pw['access_secret']

In [495]:
# Initiate session with Twitter API
auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)
api = tweepy.API(auth)

In [496]:
# Query for every tweet id in enhanced twitter archive and save tweet-information in json-format to 'tweet_json.txt'
tweet_jsons = {}
tweet_id_errors = []
start = time.time()
count = 0


with open('tweet_json.txt', 'w') as outfile:
    
    for tweet_id in df_twitter_archive_enhanced['tweet_id']:
        count +=1
        try:
            # Query API for data of tweet id, setting tweet_mode parameter to 'extended'
            tweet = api.get_status(tweet_id, tweet_mode='extended', wait_on_rate_limit=True, wait_on_rate_limit_notify=True)
            # Measure elapsed time
            mid_s = time.time()
            # Print id and time elapsed
            ## print(str(count) + ": " + str(tweet_id) + " was successfully queried at " + str(mid_s - start) )
            # Write json of tweet to 'tweet_json.txt'
            json.dump(tweet._json, outfile)
            # New line
            outfile.write("\n")

        # Not best practice to catch all exceptions but fine for this short script
        except Exception as error:
            mid_f = time.time()
            ## print(str(count) + ": " + str(tweet_id) + " could not be queried " + str(mid_f - start) + str(error))
            # Gather ids of id's without status
            tweet_id_errors.append([count, str(tweet_id)])
            
    end = time.time()
    print(end - start)
df_tweet_id_errors = pd.DataFrame(tweet_id_errors, columns=['tweet_index', 'tweet_id'])

Rate limit reached. Sleeping for: 690
Rate limit reached. Sleeping for: 710


1928.3107409477234


In [497]:
# Create empty list
tweet_api_info = []

# Open file with handle as best practice and for every json-object append interesting info to tweet_api_info
try:
    with open('tweet_json.txt') as file:
        
        for jsonObj in file:
            tweet = json.loads(jsonObj)
            tweet_api_info.append([tweet['id'], tweet['retweet_count'], tweet['favorite_count']])

# Catch exception, if file doesn't exist
except FileNotFoundError as fnf_error:
    print(fnf_error)

# Convert list to pandas dataframe
df_tweet_api_info = pd.DataFrame(tweet_api_info, columns=['tweet_id', 'retweet_count', 'favorite_count'])

In [498]:
# For development purposes store api-info to csv
df_tweet_api_info.to_csv('df_tweet_api_info.csv', encoding='utf-8')

<a id='assess'></a>
## 2. Assess data

After gathering each of the above pieces of data, they are assessed visually and programmatically for quality and tidiness issues. Requirements to be met:

- Quality requirements:
    - Completeness: All necessary records in dataframes, no specific rows, columns or cells missing.
    - Validity: No records available, that do not conform schema.
    - Accuracy: No wrong data, that is valid.
    - Consistency: No data, that is valid and accurate, but referred to in multiple correct ways.
- Tidiniss requirements (as defined by Hadley Wickham):
    - each variable is a column
    - each observation is a row
    - each type of observational unit is a table.

### a. Visual assessment

In [499]:
# Assess df_twitter_archive_enhanced visually
df_twitter_archive_enhanced.sample(n=5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
317   834786237630337024                    NaN                  NaN   
373   828376505180889089                    NaN                  NaN   
1671  682393905736888321                    NaN                  NaN   
2054  671390180817915904                    NaN                  NaN   
354   831262627380748289                    NaN                  NaN   

                      timestamp  \
317   2017-02-23 15:25:23 +0000   
373   2017-02-05 22:55:23 +0000   
1671  2015-12-31 02:52:40 +0000   
2054  2015-11-30 18:07:47 +0000   
354   2017-02-13 22:03:49 +0000   

                                                 source  \
317   <a href="http://twitter.com/download/iphone" r...   
373   <a href="http://twitter.com/download/iphone" r...   
1671  <a href="http://twitter.com/download/iphone" r...   
2054  <a href="http://twitter.com/download/iphone" r...   
354   <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
317   This is Tobi. She is properly fetching her sho...                  NaN   
373   This is Beebop. Her name means "Good Dog" in r...                  NaN   
1671  This is Kreg. He's riding an invisible jet ski...                  NaN   
2054  Striped dog here. Having fun playing on back. ...                  NaN   
354   This is Tucker. He's feeling h*ckin festive an...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
317                        NaN                        NaN   
373                        NaN                        NaN   
1671                       NaN                        NaN   
2054                       NaN                        NaN   
354                        NaN                        NaN   

                                          expanded_urls  rating_numerator  \
317   https://twitter.com/dog_rates/status/834786237...                13   
373   https://twitter.com/dog_rates/status/828376505...                13   
1671  https://twitter.com/dog_rates/status/682393905...                11   
2054  https://twitter.com/dog_rates/status/671390180...                 7   
354   https://twitter.com/dog_rates/status/831262627...                12   

      rating_denominator    name doggo floofer pupper puppo  
317                   10    Tobi  None    None   None  None  
373                   10  Beebop  None    None   None  None  
1671                  10    Kreg  None    None   None  None  
2054                  10    None  None    None   None  None  
354                   10  Tucker  None    None   None  None

In [500]:
# Assess df_image_predictions visually
df_image_predictions.sample(n=10)

tweet_id                                            jpg_url  \
338   672239279297454080    https://pbs.twimg.com/media/CVRGDrsWsAAUWSF.jpg   
717   685663452032069632  https://pbs.twimg.com/ext_tw_video_thumb/68566...   
1416  771136648247640064    https://pbs.twimg.com/media/CrOgsIBWYAA8Dtb.jpg   
1819  834209720923721728    https://pbs.twimg.com/media/C5O1UAaWIAAMBMd.jpg   
1661  811744202451197953    https://pbs.twimg.com/media/C0PlCQjXAAA9TIh.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
770   689289219123089408  https://pbs.twimg.com/ext_tw_video_thumb/68928...   
882   698710712454139905    https://pbs.twimg.com/media/CbJRrigW0AIcJ2N.jpg   
2070  891327558926688256    https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
190   669393256313184256    https://pbs.twimg.com/media/CUopnHPVEAAcL2o.jpg   

      img_num                        p1   p1_conf  p1_dog                  p2  \
338         1                       pug  0.332536    True      French_bulldog   
717         1  Chesapeake_Bay_retriever  0.171174    True         tennis_ball   
1416        1                   bathtub  0.368660   False    golden_retriever   
1819        1          golden_retriever  0.754799    True            Pekinese   
1661        1                  Pekinese  0.386082    True  Labrador_retriever   
3           1       Rhodesian_ridgeback  0.408143    True             redbone   
770         1                snowmobile  0.254642   False       assault_rifle   
882         1                   Samoyed  0.329895    True               shoji   
2070        2                    basset  0.555712    True    English_springer   
190         1            cocker_spaniel  0.359843    True    Blenheim_spaniel   

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
338   0.258124    True                 bull_mastiff  0.120873    True  
717   0.090644   False                       racket  0.048508   False  
1416  0.297402    True                          tub  0.201711   False  
1819  0.197861    True           Labrador_retriever  0.008654    True  
1661  0.202862    True             golden_retriever  0.170487    True  
3     0.360687    True           miniature_pinscher  0.222752    True  
770   0.129558   False                        rifle  0.110875   False  
882   0.165772   False                       prison  0.103596   False  
2070  0.225770    True  German_short-haired_pointer  0.175219    True  
190   0.139519    True                   toy_poodle  0.132746    True

In [501]:
df_image_predictions.query('tweet_id == 749317047558017024')['jpg_url']

1265    https://pbs.twimg.com/ext_tw_video_thumb/74931...
Name: jpg_url, dtype: object

In [502]:
# Assess df_tweet_api_info visually
df_tweet_api_info.sample(n=10)

tweet_id  retweet_count  favorite_count
1546  687732144991551489            667            1765
484   813112105746448384           2808           10546
2028  671485057807351808            227             735
614   793962221541933056           5030           17263
1425  696100768806522880            654            1938
2035  671182547775299584            324            1084
284   837110210464448512           2370           16133
1939  673350198937153538            203             728
1724  679047485189439488            661            2238
1893  674271431610523648            710            1529

### b. Programmatic assessment

In [503]:
# Check datatypes and non-null count of columns in df_twitter_archive_enhanced
df_twitter_archive_enhanced.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [504]:
# Check if column rating_denominator of df_twitter_archive_enhanced only consists of value 10
df_twitter_archive_enhanced.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

In [505]:
# Check column rating_numerator for realistic values
df_twitter_archive_enhanced.rating_numerator.value_counts() 

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

In [506]:
# Check for most common values in column name of df_twitter_archive_enhanced
df_twitter_archive_enhanced.name.value_counts() 

None         745
a             55
Charlie       12
Cooper        11
Oliver        11
            ... 
Alejandro      1
Tessa          1
Arlo           1
Ron            1
Cupid          1
Name: name, Length: 957, dtype: int64

In [645]:
# Inspect texts of some tweets without dog name 
df_twitter_archive_enhanced.query('name == "None"').sample(n=5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2184  668988183816871936                    NaN                  NaN   
798   772877495989305348                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
1714  680440374763077632                    NaN                  NaN   
788   774314403806253056                    NaN                  NaN   

                      timestamp  \
2184  2015-11-24 03:03:06 +0000   
798   2016-09-05 19:22:09 +0000   
2343  2015-11-16 01:59:36 +0000   
1714  2015-12-25 17:30:01 +0000   
788   2016-09-09 18:31:54 +0000   

                                                 source  \
2184  <a href="http://twitter.com/download/iphone" r...   
798   <a href="http://twitter.com" rel="nofollow">Tw...   
2343  <a href="http://twitter.com/download/iphone" r...   
1714  <a href="http://twitter.com/download/iphone" r...   
788   <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
2184  Honor to rate this dog. Lots of fur on him. Tw...                  NaN   
798   You need to watch these two doggos argue throu...                  NaN   
2343  Let's hope this flight isn't Malaysian (lol). ...                  NaN   
1714  Merry Christmas. My gift to you is this tiny u...                  NaN   
788   I WAS SENT THE ACTUAL DOG IN THE PROFILE PIC B...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
2184                       NaN                        NaN   
798                        NaN                        NaN   
2343                       NaN                        NaN   
1714                       NaN                        NaN   
788                        NaN                        NaN   

                                          expanded_urls  rating_numerator  \
2184  https://twitter.com/dog_rates/status/668988183...                 7   
798   https://twitter.com/dog_rates/status/772877495...                11   
2343  https://twitter.com/dog_rates/status/666073100...                10   
1714  https://twitter.com/dog_rates/status/680440374...                11   
788   https://twitter.com/dog_rates/status/774314403...                14   

      rating_denominator  name doggo floofer pupper puppo  
2184                  10  None  None    None   None  None  
798                   10  None  None    None   None  None  
2343                  10  None  None    None   None  None  
1714                  10  None  None    None   None  None  
788                   10  None  None    None   None  None

In [646]:
# Inspect texts of some tweets with value "a" in column "name"
df_twitter_archive_enhanced.query('name == "a"').sample(n=5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1737  679530280114372609                    NaN                  NaN   
1785  677644091929329666                    NaN                  NaN   
2161  669564461267722241                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2287  667177989038297088                    NaN                  NaN   

                      timestamp  \
1737  2015-12-23 05:13:38 +0000   
1785  2015-12-18 00:18:36 +0000   
2161  2015-11-25 17:13:02 +0000   
2352  2015-11-16 00:04:52 +0000   
2287  2015-11-19 03:10:02 +0000   

                                                 source  \
1737  <a href="http://twitter.com/download/iphone" r...   
1785  <a href="http://twitter.com/download/iphone" r...   
2161  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2287  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1737  Guys this really needs to stop. We've been ove...                  NaN   
1785  This is a dog swinging. I really enjoyed it so...                  NaN   
2161  This is a Coriander Baton Rouge named Alfredo....                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2287  This is a Dasani Kingfisher from Maine. His na...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1737                       NaN                        NaN   
1785                       NaN                        NaN   
2161                       NaN                        NaN   
2352                       NaN                        NaN   
2287                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
1737  https://twitter.com/dog_rates/status/679530280...                 7   
1785  https://twitter.com/dog_rates/status/677644091...                11   
2161  https://twitter.com/dog_rates/status/669564461...                10   
2352  https://twitter.com/dog_rates/status/666044226...                 6   
2287  https://twitter.com/dog_rates/status/667177989...                 8   

      rating_denominator name doggo floofer pupper puppo  
1737                  10    a  None    None   None  None  
1785                  10    a  None    None   None  None  
2161                  10    a  None    None   None  None  
2352                  10    a  None    None   None  None  
2287                  10    a  None    None   None  None

In [507]:
# Check datatypes and non-null count of columns in df_image_predictions
df_image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [508]:
# Check for percentiles and standard deviation of image prediction confidence
df_image_predictions.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [509]:
# Check if p1_conf means highest condifdence level
np.mean(df_image_predictions.p1_conf > df_image_predictions.p2_conf)

1.0

In [510]:
# Check datatypes and non-null count of columns in df_tweet_api_info
df_tweet_api_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   tweet_id        2331 non-null   int64
 1   retweet_count   2331 non-null   int64
 2   favorite_count  2331 non-null   int64
dtypes: int64(3)
memory usage: 54.8 KB


In [511]:
df_tweet_id_errors

tweet_index            tweet_id
0            20  888202515573088257
1            96  873697596434513921
2           102  872668790621863937
3           105  872261713294495745
4           119  869988702071779329
5           133  866816280283807744
6           156  861769973181624320
7           183  856602993587888130
8           212  851953902622658560
9           248  845459076796616705
10          254  844704788403113984
11          261  842892208864923648
12          297  837366284874571778
13          299  837012587749474308
14          364  829374341691346946
15          383  827228250799742977
16          507  812747805718642688
17          567  802247111496568832
18          751  779123168116150273
19          785  775096608509886464
20          816  771004394259247104
21          819  770743923962707968
22          891  759566828574212096
23          933  754011816964026368
24         1727  680055455951884288

In [512]:
# Check for duplicate column names across three dataframes
all_columns = pd.Series(list(df_twitter_archive_enhanced) + list(df_image_predictions) + list(df_tweet_api_info))
all_columns[all_columns.duplicated()]

17    tweet_id
29    tweet_id
dtype: object

In [513]:
# Statistical description of dataframe, min-values, max-values, quartiles
df_tweet_api_info.describe()

tweet_id  retweet_count  favorite_count
count  2.331000e+03    2331.000000      2331.00000
mean   7.419079e+17    2716.009438      7595.36894
std    6.823170e+16    4594.501971     11786.22056
min    6.660209e+17       1.000000         0.00000
25%    6.782670e+17     548.000000      1327.00000
50%    7.182469e+17    1275.000000      3306.00000
75%    7.986692e+17    3154.000000      9302.00000
max    8.924206e+17   78211.000000    156978.00000

### Findings, which contradict requirements:

##### `df_twitter_archive_enhanced` table

- Validity: Not all tweets are dog ratings, some are retweets. 181 Retweets, identied by values in columns `retweeted_status_id`, `retweeted_status_user_id` and `retweeted_status_timestamp`.
- Validity: Not all tweets are dog ratings, some are retweets. Retweets: 78 non-null values in columns `in_reply_to_status_id` and `in_reply_to_user_id`.
- Validity: Not all values in column `rating_denominator`are equal to 10.
- Completeness: Not all rows have a value in one of the following columns: `doggo`, `floofer`, `pupper`, `puppo`.
- Completeness: Missing values in columns `doggo`, `floofer`, `pupper`, `puppo`

- Consistency: Duplicate info in column `text` to columns `rating_numerator`, `rating_denominator`, `name`, `doggo`, `floofer`, `pupper`, `puppo`.

- Completeness: There are 745 entries 'None' and 55 entries 'a' in the column `name`. 
- Accuracy: Column `timestamp` is not of type datetime.



- Validity: In column `expanded_urls` the urls match mainly `https://twitter.com/dog_rates/status/`, yet there is one example where the url starts with `https://gofundme.com/ydvmve-surgery-for-`.


##### `df_image_predictions` table
- Consistency: Up to three different images with three possible interpretations what is depicted on the image. Ratings of interest always refer to one dog. Not all images depict dogs (regarding the image recognition/prediction table).
- Accuracy: Not all image predications are correct, the confidence level is an indicator.

##### `df_tweet_api_info` table
- ...

#### Tidiness Observations:
- Variable `text`contains multiple observational units `rating_numerator`, `rating_denominator`, `name`, `doggo`, `floofer`, `pupper`, `puppo`.
- One variable (dog stage) in four columns in `treatments` table (doggo, floofer, pupper and puppo) of type categorical. 
- One observational unit in in three tables, `df_twitter_archive_enhanced`, `df_image_predictions`and `df_tweet_api_info`, different amount of rows.

<a id='clean'></a>
## 3. Clean data

In [794]:
# Create copies for cleaning process to preserve original dataframes
df_twitter_archive_enhanced_clean = df_twitter_archive_enhanced.copy()
df_image_predictions_clean = df_image_predictions.copy()
df_tweet_api_info_clean = df_tweet_api_info.copy()

1. Clean tidiness issues.
2. Clean completeness issues.
3. Clean remaining quality issues.

### Issue 1:
#### Observe:
- Validity: Not all tweets are dog ratings, some are retweets. 181 Retweets, identied by values in columns `retweeted_status_id`, `retweeted_status_user_id` and `retweeted_status_timestamp`.

#### Define:
- Drop rows, which can be identified as retweets by non-null values in columns `retweeted_status_id`, `retweeted_status_user_id` and `retweeted_status_timestamp`. Furthermore drop columns `retweeted_status_id`, `retweeted_status_user_id` and `retweeted_status_timestamp`, as not necessary anymore.

#### Code:

In [795]:
# Keep only rows without values in 'retweeted_status_id'
df_twitter_archive_enhanced_clean = df_twitter_archive_enhanced_clean[
    df_twitter_archive_enhanced_clean['retweeted_status_id'].isnull()]

# Drop variables which are only necessary for retweets
df_twitter_archive_enhanced_clean.drop(['retweeted_status_id', 'retweeted_status_user_id', 
                                        'retweeted_status_timestamp'], axis=1, inplace=True)

#### Test:

In [796]:
# Check if amount twitter ids reduced as expected
assert len(df_twitter_archive_enhanced_clean) == len(df_twitter_archive_enhanced)-181

In [797]:
# Check if columnns `in_reply_to_status_id`, `in_reply_to_user_id` and `retweeted_status_timestamp` dropped
list(df_twitter_archive_enhanced)

['tweet_id',
 'in_reply_to_status_id',
 'in_reply_to_user_id',
 'timestamp',
 'source',
 'text',
 'retweeted_status_id',
 'retweeted_status_user_id',
 'retweeted_status_timestamp',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'name',
 'doggo',
 'floofer',
 'pupper',
 'puppo']

### Issue 2:
#### Observe:
- Validity: Not all tweets are dog ratings, some are replies. 78 replies, identied by values in columns `in_reply_to_status_id` and `in_reply_to_user_id`.

#### Define:
- Drop rows, which can be identified as replies by non-null values in columns `in_reply_to_status_id` and `in_reply_to_user_id`. Furthermore drop columns `in_reply_to_status_id` and `in_reply_to_user_id`, as not necessary anymore.

#### Code:

In [798]:
# Keep only rows without values in 'in_reply_to_status_id'
df_twitter_archive_enhanced_clean = df_twitter_archive_enhanced_clean[df_twitter_archive_enhanced_clean[
    'in_reply_to_status_id'].isnull()]

# Drop variables which are only necessary for replies
df_twitter_archive_enhanced_clean.drop(['in_reply_to_status_id', 'in_reply_to_user_id'], 
                                       axis=1, inplace=True)

#### Test:

In [799]:
# Check if amount twitter ids reduced as expected
assert len(df_twitter_archive_enhanced_clean) == len(df_twitter_archive_enhanced)-181-78

In [800]:
# Check if columnns `in_reply_to_status_id`, `in_reply_to_user_id` and `retweeted_status_timestamp` dropped
list(df_twitter_archive_enhanced_clean)

['tweet_id',
 'timestamp',
 'source',
 'text',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'name',
 'doggo',
 'floofer',
 'pupper',
 'puppo']

In [801]:
# Check if any row left without value in rating_numerator column
len(df_twitter_archive_enhanced_clean[df_twitter_archive_enhanced_clean['rating_numerator'].isnull()])

0

### Issue 3:
#### Observe:
- Validity: Not all values in column `rating_denominator`are equal to 10. Not all ratings seem to be valid.

#### Define
- Extract `rating_numerator`, `rating_denominator`variables from the text column using regular expressions and pandas' str.extract method. Drop the ratings column when done. Delete rows which do not have 10 as denominator.

#### Code:

In [802]:
# Extract rating_numerator/rating_denominator via regex
df_twitter_archive_enhanced_clean['rating'] = df_twitter_archive_enhanced_clean.text.str.extract(
    '(\d{1,2}\.?\d{1,2}?\/10)', expand=True)

# Split the values in 'rating' column to columns 'rating_numerator' and 'rating_denominator'
df_twitter_archive_enhanced_clean['rating_numerator_text'], df_twitter_archive_enhanced_clean[
    'rating_denominator_text'] = df_twitter_archive_enhanced_clean['rating'].str.split('/', 1).str


/usr/local/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:7: FutureWarning: Columnar iteration over characters will be deprecated in future releases.
  import sys


In [803]:
df_twitter_archive_enhanced_clean['rating_numerator_text'] = df_twitter_archive_enhanced_clean['rating_numerator_text'].replace(np.nan, 'None')

# Function to apply, which goes through 3 available guesses and picks the one with highest confidence beeing dog
def fill_missing_numerators(tweet):
    if tweet['rating_numerator_text'] == 'None':
        rating_numerator = tweet['rating_numerator']
        return rating_numerator
    else:
        rating_numerator = tweet['rating_numerator_text']
        return rating_numerator

# Apply function to dataframe    
df_twitter_archive_enhanced_clean['rating_numerator_merge'] = df_twitter_archive_enhanced_clean.apply(fill_missing_numerators, axis=1)


df_twitter_archive_enhanced_clean['rating_numerator'] = df_twitter_archive_enhanced_clean['rating_numerator_merge']

# Note: axis=1 denotes that referring to a column, not a row
df_twitter_archive_enhanced_clean.drop(['rating_numerator_merge', 'rating_numerator_text'], axis=1, inplace=True)


tweet_id                  timestamp  \
0     892420643555336193  2017-08-01 16:23:56 +0000   
1     892177421306343426  2017-08-01 00:17:27 +0000   
2     891815181378084864  2017-07-31 00:18:03 +0000   
3     891689557279858688  2017-07-30 15:58:51 +0000   
4     891327558926688256  2017-07-29 16:00:24 +0000   
...                  ...                        ...   
2351  666049248165822465  2015-11-16 00:24:50 +0000   
2352  666044226329800704  2015-11-16 00:04:52 +0000   
2353  666033412701032449  2015-11-15 23:21:54 +0000   
2354  666029285002620928  2015-11-15 23:05:30 +0000   
2355  666020888022790149  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  \
0     This is Phineas. He's a mystical boy. Only eve...   
1     This is Tilly. She's just checking pup on you....   
2     This is Archie. He is a rare Norwegian Pouncin...   
3     This is Darla. She commenced a snooze mid meal...   
4     This is Franklin. He would like you to stop ca...   
...                                                 ...   
2351  Here we have a 1949 1st generation vulpix. Enj...   
2352  This is a purebred Piers Morgan. Loves to Netf...   
2353  Here is a very happy pup. Big fan of well-main...   
2354  This is a western brown Mitsubishi terrier. Up...   
2355  Here we have a Japanese Irish Setter. Lost eye...   

                                          expanded_urls rating_numerator  \
0     https://twitter.com/dog_rates/status/892420643...               13   
1     https://twitter.com/dog_rates/status/892177421...               13   
2     https://twitter.com/dog_rates/status/891815181...               12   
3     https://twitter.com/dog_rates/status/891689557...               13   
4     https://twitter.com/dog_rates/status/891327558...               12   
...                                                 ...              ...   
2351  https://twitter.com/dog_rates/status/666049248...                5   
2352  https://twitter.com/dog_rates/status/666044226...                6   
2353  https://twitter.com/dog_rates/status/666033412...                9   
2354  https://twitter.com/dog_rates/status/666029285...                7   
2355  https://twitter.com/dog_rates/status/666020888...                8   

      rating_denominator      name doggo floofer pupper puppo rating  \
0                     10   Phineas  None    None   None  None  13/10   
1                     10     Tilly  None    None   None  None  13/10   
2                     10    Archie  None    None   None  None  12/10   
3                     10     Darla  None    None   None  None  13/10   
4                     10  Franklin  None    None   None  None  12/10   
...                  ...       ...   ...     ...    ...   ...    ...   
2351                  10      None  None    None   None  None    NaN   
2352                  10         a  None    None   None  None    NaN   
2353                  10         a  None    None   None  None    NaN   
2354                  10         a  None    None   None  None    NaN   
2355                  10      None  None    None   None  None    NaN   

     rating_denominator_text  
0                         10  
1                         10  
2                         10  
3                         10  
4   

In [804]:
df_twitter_archive_enhanced_clean['rating_denominator_text'] = df_twitter_archive_enhanced_clean['rating_denominator_text'].replace(np.nan, 'None')

# Function to apply, which goes through 3 available guesses and picks the one with highest confidence beeing dog
def fill_missing_numerators(tweet):
    if tweet['rating_denominator_text'] == 'None':
        rating_numerator = tweet['rating_denominator']
        return rating_numerator
    else:
        rating_numerator = tweet['rating_denominator_text']
        return rating_numerator

# Apply function to dataframe    
df_twitter_archive_enhanced_clean['rating_denominator_merge'] = df_twitter_archive_enhanced_clean.apply(fill_missing_numerators, axis=1)


df_twitter_archive_enhanced_clean['rating_denominator'] = df_twitter_archive_enhanced_clean['rating_denominator_merge']

# Note: axis=1 denotes that referring to a column, not a row
df_twitter_archive_enhanced_clean.drop(['rating_denominator_merge', 'rating_denominator_text'], axis=1, inplace=True)

tweet_id                  timestamp  \
0     892420643555336193  2017-08-01 16:23:56 +0000   
1     892177421306343426  2017-08-01 00:17:27 +0000   
2     891815181378084864  2017-07-31 00:18:03 +0000   
3     891689557279858688  2017-07-30 15:58:51 +0000   
4     891327558926688256  2017-07-29 16:00:24 +0000   
...                  ...                        ...   
2351  666049248165822465  2015-11-16 00:24:50 +0000   
2352  666044226329800704  2015-11-16 00:04:52 +0000   
2353  666033412701032449  2015-11-15 23:21:54 +0000   
2354  666029285002620928  2015-11-15 23:05:30 +0000   
2355  666020888022790149  2015-11-15 22:32:08 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  \
0     This is Phineas. He's a mystical boy. Only eve...   
1     This is Tilly. She's just checking pup on you....   
2     This is Archie. He is a rare Norwegian Pouncin...   
3     This is Darla. She commenced a snooze mid meal...   
4     This is Franklin. He would like you to stop ca...   
...                                                 ...   
2351  Here we have a 1949 1st generation vulpix. Enj...   
2352  This is a purebred Piers Morgan. Loves to Netf...   
2353  Here is a very happy pup. Big fan of well-main...   
2354  This is a western brown Mitsubishi terrier. Up...   
2355  Here we have a Japanese Irish Setter. Lost eye...   

                                          expanded_urls rating_numerator  \
0     https://twitter.com/dog_rates/status/892420643...               13   
1     https://twitter.com/dog_rates/status/892177421...               13   
2     https://twitter.com/dog_rates/status/891815181...               12   
3     https://twitter.com/dog_rates/status/891689557...               13   
4     https://twitter.com/dog_rates/status/891327558...               12   
...                                                 ...              ...   
2351  https://twitter.com/dog_rates/status/666049248...                5   
2352  https://twitter.com/dog_rates/status/666044226...                6   
2353  https://twitter.com/dog_rates/status/666033412...                9   
2354  https://twitter.com/dog_rates/status/666029285...                7   
2355  https://twitter.com/dog_rates/status/666020888...                8   

     rating_denominator      name doggo floofer pupper puppo rating  
0                    10   Phineas  None    None   None  None  13/10  
1                    10     Tilly  None    None   None  None  13/10  
2                    10    Archie  None    None   None  None  12/10  
3                    10     Darla  None    None   None  None  13/10  
4                    10  Franklin  None    None   None  None  12/10  
...                 ...       ...   ...     ...    ...   ...    ...  
2351                 10      None  None    None   None  None    NaN  
2352                 10         a  None    None   None  None    NaN  
2353                 10         a  None    None   None  None    NaN  
2354                 10         a  None    None   None  None    NaN  
2355                 10      None  None    None   None  None    NaN  

[2097 rows x 13 columns]

In [805]:

# Drop rows which do not conform the (\d{1,2}\/10) format for ratings
df_twitter_archive_enhanced_clean = df_twitter_archive_enhanced_clean[df_twitter_archive_enhanced_clean.rating_numerator.notnull()]
df_twitter_archive_enhanced_clean = df_twitter_archive_enhanced_clean[df_twitter_archive_enhanced_clean.rating_denominator.notnull()]

# Convert columns to int
df_twitter_archive_enhanced_clean.rating_numerator = pd.to_numeric(df_twitter_archive_enhanced_clean.rating_numerator)
df_twitter_archive_enhanced_clean.rating_denominator = pd.to_numeric(df_twitter_archive_enhanced_clean.rating_denominator)

# Note: axis=1 denotes that referring to a column, not a row
df_twitter_archive_enhanced_clean.drop('rating', axis=1, inplace=True)

In [809]:
df_twitter_archive_enhanced_clean = df_twitter_archive_enhanced_clean.query('rating_denominator == 10')

#### Test:

In [810]:
# Check if columns `rating_numerator` and `rating_denominator` have int64 as datatype
df_twitter_archive_enhanced_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2084 entries, 0 to 2355
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   tweet_id            2084 non-null   int64  
 1   timestamp           2084 non-null   object 
 2   source              2084 non-null   object 
 3   text                2084 non-null   object 
 4   expanded_urls       2081 non-null   object 
 5   rating_numerator    2084 non-null   float64
 6   rating_denominator  2084 non-null   int64  
 7   name                2084 non-null   object 
 8   doggo               2084 non-null   object 
 9   floofer             2084 non-null   object 
 10  pupper              2084 non-null   object 
 11  puppo               2084 non-null   object 
dtypes: float64(1), int64(2), object(9)
memory usage: 211.7+ KB


In [811]:
# Check for values in rating_numerator
df_twitter_archive_enhanced_clean.rating_numerator.value_counts()

12.00      486
10.00      438
11.00      415
13.00      289
9.00       152
8.00        97
7.00        51
14.00       39
5.00        32
6.00        32
3.00        19
4.00        14
2.00         9
1.00         4
9.75         1
0.00         1
11.26        1
11.27        1
13.50        1
420.00       1
1776.00      1
Name: rating_numerator, dtype: int64

In [812]:
# Check for values in rating_numerator
df_twitter_archive_enhanced_clean.rating_denominator.value_counts()

10    2084
Name: rating_denominator, dtype: int64

### Issue 4:
#### Observe:
- Completeness: Missing values in columns `doggo`, `floofer`, `pupper`, `puppo`

#### Define: 
- Extract `dogger`, `floofer`, `pupper` and `puppo` variables from the text column using regular expressions and pandas' str.extract method.

In [713]:
# Extract stage via regex
df_twitter_archive_enhanced_clean['stage'] = df_twitter_archive_enhanced_clean.text.str.extract(
    '(puppo|pupper|floofer|doggo)', expand=True)


In [714]:
df_twitter_archive_enhanced_clean.stage.value_counts()

pupper     192
doggo       73
puppo       28
floofer      4
Name: stage, dtype: int64

### Issue 5:
#### Observe:
- Tidiness: One variable (dog stage) in four columns in `treatments` table (doggo, floofer, pupper and puppo), all four not of type categorical. 

#### Define:
- Melt the variables *doggo*, *floofer*, *pupper* and *puppo* columns to a *stage* wit the [melt function](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html). Extract missing stage information from `text` column. Convert new column `stage` to categorical.

#### Code:

In [667]:
list(df_twitter_archive_enhanced_clean)

['tweet_id',
 'timestamp',
 'source',
 'text',
 'expanded_urls',
 'rating_numerator',
 'rating_denominator',
 'name',
 'doggo',
 'floofer',
 'pupper',
 'puppo']

In [668]:
len(df_twitter_archive_enhanced_clean)

1673

In [669]:
# Replace None with NaN
for stage in stages: 
    df_twitter_archive_enhanced_clean[stage] = df_twitter_archive_enhanced_clean[stage].replace('None', np.nan)

# Use melt function to melt pupper, doggo, puppo, floofer to new column 'stage
df_twitter_archive_enhanced_clean = pd.melt(df_twitter_archive_enhanced_clean, 
                                            id_vars=['tweet_id', 'timestamp', 'source', 'text', 
                                                     'expanded_urls', 'rating_numerator', 'rating_denominator', 
                                                     'name'], var_name='var_name', value_name='stage')


In [670]:
# Drop variables x
#df_twitter_archive_enhanced_clean.drop('var_name', axis=1, inplace=True)

# Keep only observations with defined stage
#df_twitter_archive_enhanced_clean = df_twitter_archive_enhanced_clean.query('stage == "pupper" or stage == "doggo" or stage == "puppo" or stage == "floofer"')

#### Test:

In [671]:
len(df_twitter_archive_enhanced_clean)

6692

In [672]:
df_twitter_archive_enhanced_clean.stage.value_counts()

           6389
pupper      191
doggo        79
puppo        23
floofer      10
Name: stage, dtype: int64

In [715]:
df_twitter_archive_enhanced_clean.head(20)

tweet_id                  timestamp  \
0   892420643555336193  2017-08-01 16:23:56 +0000   
1   892177421306343426  2017-08-01 00:17:27 +0000   
2   891815181378084864  2017-07-31 00:18:03 +0000   
3   891689557279858688  2017-07-30 15:58:51 +0000   
4   891327558926688256  2017-07-29 16:00:24 +0000   
5   891087950875897856  2017-07-29 00:08:17 +0000   
6   890971913173991426  2017-07-28 16:27:12 +0000   
7   890729181411237888  2017-07-28 00:22:40 +0000   
8   890609185150312448  2017-07-27 16:25:51 +0000   
9   890240255349198849  2017-07-26 15:59:51 +0000   
10  890006608113172480  2017-07-26 00:31:25 +0000   
11  889880896479866881  2017-07-25 16:11:53 +0000   
12  889665388333682689  2017-07-25 01:55:32 +0000   
13  889638837579907072  2017-07-25 00:10:02 +0000   
14  889531135344209921  2017-07-24 17:02:04 +0000   
15  889278841981685760  2017-07-24 00:19:32 +0000   
16  888917238123831296  2017-07-23 00:22:39 +0000   
17  888804989199671297  2017-07-22 16:56:37 +0000   
18  888554962724278272  2017-07-22 00:23:06 +0000   
20  888078434458587136  2017-07-20 16:49:33 +0000   

                                               source  \
0   <a href="http://twitter.com/download/iphone" r...   
1   <a href="http://twitter.com/download/iphone" r...   
2   <a href="http://twitter.com/download/iphone" r...   
3   <a href="http://twitter.com/download/iphone" r...   
4   <a href="http://twitter.com/download/iphone" r...   
5   <a href="http://twitter.com/download/iphone" r...   
6   <a href="http://twitter.com/download/iphone" r...   
7   <a href="http://twitter.com/download/iphone" r...   
8   <a href="http://twitter.com/download/iphone" r...   
9   <a href="http://twitter.com/download/iphone" r...   
10  <a href="http://twitter.com/download/iphone" r...   
11  <a href="http://twitter.com/download/iphone" r...   
12  <a href="http://twitter.com/download/iphone" r...   
13  <a href="http://twitter.com/download/iphone" r...   
14  <a href="http://twitter.com/download/iphone" r...   
15  <a href="http://twitter.com/download/iphone" r...   
16  <a href="http://twitter.com/download/iphone" r...   
17  <a href="http://twitter.com/download/iphone" r...   
18  <a href="http://twitter.com/download/iphone" r...   
20  <a href="http://twitter.com/download/iphone" r...   

                                                 text  \
0   This is Phineas. He's a mystical boy. Only eve...   
1   This is Tilly. She's just checking pup on you....   
2   This is Archie. He is a rare Norwegian Pouncin...   
3   This is Darla. She commenced a snooze mid meal...   
4   This is Franklin. He would like you to stop ca...   
5   Here we have a majestic great white breaching ...   
6   Meet Jax. He enjoys ice cream so much he gets ...   
7   When you watch your owner call another dog a g...   
8   This is Zoey. She doesn't want to be one of th...   
9   This is Cassie. She is a college pup. Studying...   
10  This is Koda. He is a South Australian decksha...   
11  This is Bruno. He is a service shark. Only get...   
12  Here's a puppo that seems to be on the fence a...   
13  This is Ted. He does his best. Sometimes that'...   
14  This is Stuart. He's sporting his favorite fan...   
15  This is Oliver. You're witnessing one of his m...   
16  This is Jim. He found a fren. Taught him how t...   
17  This is Zeke. He has a new stick. Very proud o...   
18  This is Ralphus. He's powering up. Attempting ...   
20  This is Gerald. He was just told he didn't get...   

                                        expanded_urls  rating_numerator  \
0   https://twitter.com/dog_rates/status/892420643...              13.0   
1   https://twitter.com/dog_rates/status/892177421...              13.0   
2   https://twitter.com/dog_rates/status/891815181...              12.0   
3   https://twitter.com/dog_rates/status/891689557...              13.0   
4   https://twitter.com/dog_rates/status/891327558...              12.0   
5   https://twitter.com/dog_rates/status/891087950... 

### Issue 6:
#### Observe:
- Consistency: Up to three different images with three possible interpretations what is depicted on the image. Ratings of interest always refer to one dog.

#### Define:
- Create function to pick image with highest confidence and drop other images. Keep results in resulting columns `race_prediction` and `race_prediction_conf`.

#### Code

In [675]:
# Is there a more elegant way to do this???

# Function to apply, which goes through 3 available guesses and picks the one with highest confidence beeing dog
def find_best_valid_prediction(tweet):
    if tweet['p1_dog'] == True:
        race_prediction = tweet['p1']
        return race_prediction
    elif tweet['p1_dog'] == False and tweet['p2_dog'] == True:
        race_prediction = tweet['p2']
        return race_prediction
    elif tweet['p1_dog'] == False and tweet['p2_dog'] == False and tweet['p3_dog'] == True:
        race_prediction = tweet['p3']
        return race_prediction
    else:
        race_prediction = 'None'
        return race_prediction

# Apply function to dataframe    
df_image_predictions_clean['race_prediction'] = df_image_predictions_clean.apply(find_best_valid_prediction, axis=1)


# Function to apply, which adds confidence level to best guess
def add_conf_best_valid_guess(tweet):
    if tweet['race_prediction'] == tweet['p1']:
        race_prediction_conf = tweet['p1_conf']
        return race_prediction_conf
    elif tweet['race_prediction'] != tweet['p1'] and tweet['race_prediction'] == tweet['p2']:
        race_prediction_conf = tweet['p2_conf']
        return race_prediction_conf
    elif tweet['race_prediction'] != tweet['p1'] and tweet['race_prediction'] != tweet['p2'] and tweet['race_prediction'] == tweet['p3']:
        race_prediction_conf = tweet['p3_conf']
        return race_prediction_conf
    else:
        race_prediction_conf = 0
        return race_prediction_conf

# Apply function to dataframe    
df_image_predictions_clean['race_prediction_conf'] = df_image_predictions_clean.apply(add_conf_best_valid_guess, axis=1)


# Drop variables which contain other guesses
df_image_predictions_clean.drop(['p1', 'p1_conf', 'p1_dog', 'p2', 'p2_conf', 'p2_dog', 'p3', 'p3_conf',
                                        'p3_dog'], axis=1, inplace=True)

#### Test:

In [676]:
# Check remaining columns
list(df_image_predictions_clean)

['tweet_id', 'jpg_url', 'img_num', 'race_prediction', 'race_prediction_conf']

In [677]:
df_image_predictions_clean.sample(n=3)

tweet_id                                          jpg_url  \
861  697242256848379904  https://pbs.twimg.com/media/Ca0aIR9WcAAHiPy.jpg   
940  703774238772166656  https://pbs.twimg.com/media/CcRO8FmW4AAzazk.jpg   
96   667724302356258817  https://pbs.twimg.com/media/CUQ7tv3W4AA3KlI.jpg   

     img_num     race_prediction  race_prediction_conf  
861        1          Eskimo_dog              0.063533  
940        1  Labrador_retriever              0.990119  
96         1                None              0.000000

### Issue 7:
#### Observe: 


#### Define:
- . Simply calculate the rating by dividing the numerator by the denominator. No matter the values, as long as it is in the range from 0 to 1.4 you are good. Just drop the outliers that are outside of that. You obviously create a new column called rating and drop the numerator and denominator post that calculation.


The first thing you always need to do is ask yourself is "how important is it for me to rescue the data in the row"? To anwer that run a boxplot of the column. Based on that do a df.query (for example) or a pd.cut that shows you the rows that are above a certain cutoff. Essentially you are asking how big of a challenge do I have if I simply drop those rows?
In this project, there are only a few outliers (numerator). One is the 1776 outlier. You should find that picture. It is hilarious! Generally speaking on outliers the conversation is how big of a deal is it to just drop them? If it does not impact your analysis I would drop them. That would also be an OK answer on this project.

tweet_id      img_num  race_prediction_conf
count  2.075000e+03  2075.000000           2075.000000
mean   7.384514e+17     1.203855              0.463269
std    6.785203e+16     0.561875              0.339216
min    6.660209e+17     1.000000              0.000000
25%    6.764835e+17     1.000000              0.139761
50%    7.119988e+17     1.000000              0.457117
75%    7.932034e+17     1.000000              0.772790
max    8.924206e+17     4.000000              0.999956

519

#### Test:

### Issue xx:
#### Observe:

- Tidiness: One observational unit in in three tables, `df_twitter_archive_enhanced`, `df_image_predictions` and `df_tweet_api_info`, different amount of rows.

#### Define
- ...

In [682]:
# Merge the two dataframes on columns 'given_name', 'surname'
df_twitter_archive_enhanced_clean = pd.merge(df_twitter_archive_enhanced_clean, df_image_predictions_clean, 
                            on='tweet_id', how='inner')

df_twitter_archive_enhanced_clean = pd.merge(df_twitter_archive_enhanced_clean, df_tweet_api_info_clean, 
                            on='tweet_id', how='inner')


In [683]:
df_twitter_archive_enhanced_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1756 entries, 0 to 1755
Data columns (total 16 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   tweet_id              1756 non-null   int64  
 1   timestamp             1756 non-null   object 
 2   source                1756 non-null   object 
 3   text                  1756 non-null   object 
 4   expanded_urls         1756 non-null   object 
 5   rating_numerator      1756 non-null   float64
 6   rating_denominator    1756 non-null   int64  
 7   name                  1756 non-null   object 
 8   var_name              1756 non-null   object 
 9   stage                 1756 non-null   object 
 10  jpg_url               1756 non-null   object 
 11  img_num               1756 non-null   int64  
 12  race_prediction       1756 non-null   object 
 13  race_prediction_conf  1756 non-null   float64
 14  retweet_count         1756 non-null   int64  
 15  favorite_count       

In [684]:
df_twitter_archive_enhanced_clean

tweet_id                  timestamp  \
0     890006608113172480  2017-07-26 00:31:25 +0000   
1     890006608113172480  2017-07-26 00:31:25 +0000   
2     890006608113172480  2017-07-26 00:31:25 +0000   
3     890006608113172480  2017-07-26 00:31:25 +0000   
4     889665388333682689  2017-07-25 01:55:32 +0000   
...                  ...                        ...   
1751  666345417576210432  2015-11-16 20:01:42 +0000   
1752  666063827256086533  2015-11-16 01:22:45 +0000   
1753  666063827256086533  2015-11-16 01:22:45 +0000   
1754  666063827256086533  2015-11-16 01:22:45 +0000   
1755  666063827256086533  2015-11-16 01:22:45 +0000   

                                                 source  \
0     <a href="http://twitter.com/download/iphone" r...   
1     <a href="http://twitter.com/download/iphone" r...   
2     <a href="http://twitter.com/download/iphone" r...   
3     <a href="http://twitter.com/download/iphone" r...   
4     <a href="http://twitter.com/download/iphone" r...   
...                                                 ...   
1751  <a href="http://twitter.com/download/iphone" r...   
1752  <a href="http://twitter.com/download/iphone" r...   
1753  <a href="http://twitter.com/download/iphone" r...   
1754  <a href="http://twitter.com/download/iphone" r...   
1755  <a href="http://twitter.com/download/iphone" r...   

                                                   text  \
0     This is Koda. He is a South Australian decksha...   
1     This is Koda. He is a South Australian decksha...   
2     This is Koda. He is a South Australian decksha...   
3     This is Koda. He is a South Australian decksha...   
4     Here's a puppo that seems to be on the fence a...   
...                                                 ...   
1751  Look at this jokester thinking seat belt laws ...   
1752  This is the happiest dog you will ever see. Ve...   
1753  This is the happiest dog you will ever see. Ve...   
1754  This is the happiest dog you will ever see. Ve...   
1755  This is the happiest dog you will ever see. Ve...   

                                          expanded_urls  rating_numerator  \
0     https://twitter.com/dog_rates/status/890006608...              13.0   
1     https://twitter.com/dog_rates/status/890006608...              13.0   
2     https://twitter.com/dog_rates/status/890006608...              13.0   
3     https://twitter.com/dog_rates/status/890006608...              13.0   
4     https://twitter.com/dog_rates/status/889665388...              13.0   
...                                                 ...               ...   
1751  https://twitter.com/dog_rates/status/666345417...              10.0   
1752  https://twitter.com/dog_rates/status/666063827...              10.0   
1753  https://twitter.com/dog_rates/status/666063827...              10.0   
1754  https://twitter.com/dog_rates/status/666063827...              10.0   
1755  https://twitter.com/dog_rates/status/666063827...              10.0   

      rating_denominator  name var_name stage  \
0                     10  Koda    doggo         
1                     10  Koda  floofer         
2                     10  Koda   pupper         
3                     10  Koda    puppo         
4                     10  None    doggo         
...                  ...   ...      ...   ...   
1751                  10  None    puppo         
1752                  10   the    doggo         
1753                  10   the  floofer         
1754                  10   the   pupper         
1755                  10   the    puppo         

                                              jpg_url  img_num  \
0     https://pbs.twimg.com/media/DFnwSY4WAAAMliS.jpg        1   
1     https://pbs.twimg.com/media/DFnwSY4WAAAMliS.jpg        1   
2     https://pbs.twimg.com/media/DFnwSY4WAAAMliS.jpg        1   
3     https://pbs.twimg.com/media/DFnwSY4WAAAMliS.jpg        1   
4     https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg        1   
...                        

#### Test

### Issue x: 
#### Observe: 
- ...

#### Define
- ....

#### Code

#### Test

Store date to twitter_archive_master.csv

<a id='analysis'></a>
## 4. Analysis & Visualization

In [ ]:
# How did dog ratings change over time

In [ ]:
# correlation dog stage to rating

In [ ]:
# most common name

In [19]:
from subprocess import call
call(['python', '-m', 'nbconvert', 'wrangle_act.ipynb'])

1

In [20]:
data

NameError: name 'data' is not defined